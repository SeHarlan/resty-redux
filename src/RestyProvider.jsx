import React, { createContext, useContext, useEffect, useReducer } from 'react';
import PropTypes from 'prop-types';

import { customFetch } from './services/custom';

const RestyContext = createContext();

const initialState = {
  selectedOption: 'GET',
  json: '',
  url: '',
  name: '',
  password: '',
  bearerToken: '',
  history: [],
  fetchData: null,
  result: []
};
function reducer(state, action) {
  switch(action.type) {
    case 'setSelectedOption':
      return { ...state, selectedOption: action.payload };
    case 'setJson':
      return { ...state, json: action.payload };
    case 'setUrl': 
      return { ...state, url: action.payload };
    case 'setName':
      return { ...state, name: action.payload };
    case 'setPassword':
      return { ...state, password: action.payload };
    case 'setBearerToken':
      return { ...state, bearerToken: action.payload };
    case 'setHistory':
      return { ...state, history: [...state.history, action.payload] };
    case 'setFetchData': 
      return { ...state, fetchData: action.payload };
    case 'setResult':
      return { ...state, result: action.payload };
    default:
      return state;
  }
}

const RestyProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  const { fetchData } = state;

  useEffect(() => {
    if(!fetchData) return; 
    dispatch({ type: 'setResult', payload: ['...Loading (or Nothing Found)'] });
    customFetch(fetchData).then(res => {
      if(res === []) {
        dispatch({ type: 'setResult', payload: ['No Results Found'] });
      } else {
        dispatch({ type: 'setResult', payload: res });
        dispatch({ type: 'setHistory', payload: fetchData });
      }
    });
  }, [fetchData]);

  return (
    <RestyContext.Provider value={{ state, dispatch }}>
      {children}
    </RestyContext.Provider>
  );
};

RestyProvider.propTypes = {
  children: PropTypes.node
};

export default RestyProvider;

export const useGlobalState = () => {
  const { state } = useContext(RestyContext);
  return state;
};
export const useDispatch = () => {
  const { dispatch } = useContext(RestyContext);
  return dispatch;
};

export const useHandleSubmit = (e) => {
  e.preventDefault();
  const { url, selectedOption, json, name, password, bearerToken } = useGlobalState();
  const dispatch = useDispatch();
  dispatch({ type: 'setFetchData', payload: ({
    url: url,
    method: selectedOption,
    json: json,
    username: name,
    password: password,
    bearerToken: bearerToken
  })
  });
};
export const useHandleChange = ({ target }, toChange) => {
  const dispatch = useDispatch();
  dispatch({ type: toChange, payload: target.value });
};

