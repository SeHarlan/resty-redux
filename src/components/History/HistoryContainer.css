.HistoryContainer {
  background-color: seashell;
  padding: 1rem;
  h3 {
    text-align: center;
    text-decoration: underline;
  }
  ul {
    padding: 0;
    margin: 0;
  }
  li {
    background-color: burlywood;
    list-style: none;
    padding: 0.5rem;
    margin: 0.5rem;
    border-radius: .6rem;
    h2 {
      font-size: 1rem;
      margin: 0;
    }
    p {
      font-size: 0.8rem;
      margin: 0;
    }
  }
}