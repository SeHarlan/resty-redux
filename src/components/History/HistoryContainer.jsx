import React from 'react';
import PropTypes from 'prop-types';
import withList from '../../HOCs/WithList';

import style from './HistoryContainer.css';
import { useGlobalState } from '../../RestyProvider';

const HistoryItem = ({ url, method }) => (
  <>
    <h2>{method}</h2>
    <p>{url}</p>
  </>
);

HistoryItem.propTypes = {
  method: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired
};

const HistoryElements = withList(HistoryItem);

const HistoryContainer = () => {
  const { history } = useGlobalState();
  return (
    <section className={style.HistoryContainer}>
      <h3>History</h3>
      <HistoryElements list={history} />
    </section>
  );

};

export default HistoryContainer;
