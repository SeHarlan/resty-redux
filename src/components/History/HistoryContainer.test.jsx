import React from 'react';
import { shallow } from 'enzyme';
import HistoryContainer from './HistoryContainer';
import RestyProvider from '../../RestyProvider';

describe('HistoryContainer component', () => {
  it('renders HistoryContainer', () => {
    const wrapper = shallow(
      <RestyProvider>
        <HistoryContainer />
      </RestyProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
