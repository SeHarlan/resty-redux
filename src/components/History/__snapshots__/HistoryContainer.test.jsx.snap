// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryContainer component renders HistoryContainer 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "bearerToken": "",
        "fetchData": null,
        "history": Array [],
        "json": "",
        "name": "",
        "password": "",
        "result": Array [],
        "selectedOption": "GET",
        "url": "",
      },
    }
  }
>
  <HistoryContainer />
</ContextProvider>
`;
